### ISSUE TITLE
Google test question

### ISSUE THREAD

**Issue by CaitlinA** [DrylandEcology/SOILWAT2#206]
Has anything changed about the way we connect to or use the google test submodule?

Previously, I have cloned and/or checked out any branch on SOILWAT2 and google test would work after the 'make test' command.

I just clean cloned SOILWAT2 and 'make test' is not working. I used the command 'git submodule init' and 'git submodule update --remote'. I received this error:

> MacBook-Pro-4:SOILWAT2 candrews$ make test
c++  -Wall -Wextra -Werror -std=gnu++11		 -isystem googletest/googletest/include -Igoogletest/googletest \
			-pthread -c googletest/googletest/src/gtest-all.cc
In file included from googletest/googletest/src/gtest-all.cc:42:
googletest/googletest/src/gtest.cc:393:12: error: missing field 'owner_' initializer [-Werror,-Wmissing-field-initializers]
GTEST_API_ GTEST_DEFINE_STATIC_MUTEX_(g_linked_ptr_mutex);
           ^
googletest/googletest/include/gtest/internal/gtest-port.h:2096:80: note: expanded from macro 'GTEST_DEFINE_STATIC_MUTEX_'
     ::testing::internal::MutexBase mutex = { PTHREAD_MUTEX_INITIALIZER, false }
                                                                               ^
1 error generated.
make: *** [libgtest.a] Error 1

Maybe I am forgetting something. I can't find anything in the repository documentation.

**Comment by dschlaep** [comment]
This is weird; I don't get this locally or see it on the CIs. We updated the submodule `googletest` a few days ago (commit https://github.com/DrylandEcology/SOILWAT2/commit/c159ebe9413b29aa6747b917cff63ec3166d4fdb): it was updated to commit https://github.com/google/googletest/commit/ba96d0b1161f540656efdaed035b3c062b60e006

Can you please check that you have the correct commit and a clean submodule by running
```
cd googletest/
git status
```
and making sure that you only get
```
HEAD detached at ba96d0b
nothing to commit, working tree clean
```

**Comment by dschlaep** [comment]
@CaitlinA What is the status on this issue?

**Comment by CaitlinA** [comment]
@dschlaep 

The googletest folder is pointing to the correct place

> MacBook-Pro-4:SOILWAT2 candrews$ cd googletest/
MacBook-Pro-4:googletest candrews$ git status
HEAD detached at ba96d0b
nothing to commit, working tree clean

I am still unable to run units tests and get the same error as above.

**Comment by dschlaep** [comment]
Hm, too bad. I cannot replicate and thus cannot debug very well. However, I found an issue description on `googletest` https://github.com/google/googletest/issues/1521 which appears to be the same as yours.

Another project on github addresses this issue temporarily (https://github.com/advancedtelematic/aktualizr/pull/841) by checking out one commit before this problem was introduced with commit google/googletest@87a4cdd, i.e, https://github.com/google/googletest/commit/7888184f28509dba839e3683409443e0b5bb8948 (see branch https://github.com/DrylandEcology/SOILWAT2/tree/bugfix_206 and commit https://github.com/DrylandEcology/SOILWAT2/commit/5baeec0f4321273b516024b962e394cd181b0be0). Does this work for you?

**Comment by CaitlinA** [comment]
The problem persists with commit 5baeec0

Additionally, I manually applied the suggested edit in google/googletest#1521 and I now get a different error:

> MacBook-Pro-4:SOILWAT2 candrews$ make test
c++  -Wall -Wextra -Werror -std=gnu++11		 -isystem googletest/googletest/include -Igoogletest/googletest \
			-pthread -c googletest/googletest/src/gtest-all.cc
ar -r libgtest.a gtest-all.o
ar: creating archive libgtest.a
c++ -Wall -Wextra -Werror  -g -O0 -DSWDEBUG -fstack-protector-all -std=gnu++11		 -D_FORTIFY_SOURCE=2 -fsanitize=undefined -fsanitize=address -c SW_Main_lib.c SW_VegEstab.c SW_Control.c generic.c rands.c Times.c mymemory.c filefuncs.c SW_Files.c SW_Model.c SW_Site.c SW_SoilWater.c SW_Markov.c SW_Weather.c SW_Sky.c SW_Output_mock.c SW_VegProd.c SW_Flow_lib.c SW_Flow.c SW_Carbon.c
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
clang: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated
make: *** [libSOILWAT2++.a] Error 1

**Comment by dschlaep** [comment]
- I now figured out how to replicate that error; by using `clang++` as the compiler, i.e.,
```
CXX=clang++ make cleaner test test_run
```
however, commit 5baeec0 resolved this error.

- Weird that the commit 5baeec0 didn't resolve it for you, but when you manually applied the change then it solved it -- weird because the commit is supposed to do just that. Are you sure that checking out the commit brought along the correct commit of `googletest` in your local copy? But it confirms that this was the problem.

- One recommendation to address the C++ mode error that you are now seeing is to separately compile C files with a c compiler and C++ files with a c++ compiler and then link them in a second step (e.g., https://stackoverflow.com/questions/19644816/linking-c-from-c-in-os-x-mavericks). You could have Nathan to try this out by re-writing the makefile locally; our makefile is almost set up like that except that we compile the SOILWAT2 library for the test binary with the c++ compiler instead the c compiler.

- You could switch to the gnu c++ compiler, e.g, 
```
CXX=g++ make cleaner test test_run
```

- You could turn off converting the errors into warnings in your local clone, i.e., remove ` -Werror` from the `makefile` (line 36).

**Comment by dschlaep** [comment]
@CaitlinA Ok, I took the opportunity to clean up the `makefile` and with commit 87aea18 you should be able to compile without changing anything (as suggested here https://github.com/DrylandEcology/SOILWAT2/issues/206#issuecomment-403599066). You will see the warnings, but they are no longer converted to errors (this is now a separate issue #208). I will merge this branch into master and close this issue once you confirm that this is resolved.

**Comment by dschlaep** [comment]
Which version of clang are you using? You may have an outdated version: this was a bug in clang that was fixed in 2016 (https://reviews.llvm.org/D24990 as explained here https://marc.info/?l=cfe-dev&m=147517413909285&w=2), but Apple has been shipping outdated clang versions, see https://stackoverflow.com/questions/40284307/undefined-symbols-when-using-clangs-undefined-sanitizer -- apparently, you need at least the toolchain from xcode 9.0 (https://developer.apple.com/documentation/code_diagnostics/undefined_behavior_sanitizer) for the `-fsanitize=undefined` to work with clang.

--> use target `testci` instead of `test` -- the latter uses sanitizers which apparently don't work on your setup which the former doesn't use, i.e., 
```
make clean testci test_run
```

instead of
```
make clean test_run
```

**Comment by CaitlinA** [comment]
Running 

`make clean testci test_run`

did allow for the tests to build and run when I was receiving the same errors as above. Luckily, updating my Command Line Tools for Xcode 9.2 has allowed me to build run tests with the normal

`make clean test_run`

I think we can close this issue and merge your pull request for branch [bugfix_206](https://github.com/DrylandEcology/SOILWAT2/pull/209).

**Comment by dschlaep** [comment]
closed by PR #209
